src/core_plugins/kibana/public/context/app.js

```diff
@@ -0,0 +1,103 @@
+import _ from 'lodash';
+
+import uiModules from 'ui/modules';
+import contextAppTemplate from './app.html';
+import './components/loading_button';
+import './components/size_picker/size_picker';
+import { bindSelectors } from './utils/selectors';
+import {
+  QueryParameterActionsProvider,
+  QUERY_PARAMETER_KEYS,
+} from './query_parameters';
+import {
+  QueryActionsProvider,
+  selectIsLoadingAnchorRow,
+  selectIsLoadingPredecessorRows,
+  selectIsLoadingSuccessorRows,
+  selectRows,
+} from './query';
+
+const module = uiModules.get('apps/context', [
+  'kibana',
+  'ngRoute',
+]);
+
+module.directive('contextApp', function ContextApp() {
+  return {
+    bindToController: true,
+    controller: ContextAppController,
+    controllerAs: 'contextApp',
+    restrict: 'E',
+    scope: {
+      anchorUid: '=',
+      columns: '=',
+      indexPattern: '=',
+      predecessorCount: '=',
+      successorCount: '=',
+      sort: '=',
+    },
+    template: contextAppTemplate,
+  };
+});
+
+function ContextAppController($scope, Private) {
+  const queryParameterActions = Private(QueryParameterActionsProvider);
+  const queryActions = Private(QueryActionsProvider);
+
+  this.state = createInitialState();
+
+  this.actions = _.mapValues({
+    ...queryParameterActions,
+    ...queryActions,
+  }, (action) => (...args) => action(this.state)(...args));
+
+  this.derivedState = bindSelectors({
+    isLoadingAnchorRow: selectIsLoadingAnchorRow,
+    isLoadingPredecessorRows: selectIsLoadingPredecessorRows,
+    isLoadingSuccessorRows: selectIsLoadingSuccessorRows,
+    rows: selectRows,
+  }, () => this.state);
+
+  /**
+   * Sync query parameters to arguments
+   */
+  $scope.$watchCollection(
+    () => _.pick(this, QUERY_PARAMETER_KEYS),
+    (newValues) => {
+      // break the watch cycle
```

Comment:
A regular `$watch` would compare the new and the old values of `this`, but what we actually want to compare here is a subset of the new properties of `this` and what is stored in `this.state.queryParameters`. I don't see how that can be done with a single `$watch`.